qiskit_experiments/test/t2hahn_backend.py

```diff
@@ -13,337 +13,206 @@
+from qiskit_aer.noise import NoiseModel, ReadoutError, RelaxationNoisePass, reset_error
+
 from qiskit_experiments.test.utils import FakeJob
 
-# Fix seed for simulations
-SEED = 9000
 
+# These pylint directives should be removed when the version of pylint is
+# updated from 2.7.1
+# pylint: disable=isinstance-second-argument-not-valid-type
+# pylint: disable=arguments-differ
+
+
+class ResetQubits(TransformationPass):
+    """Pass to inject reset instructions for each qubit
+
+    The resets are used to add qubit initialization error.
+    """
+
+    def run(self, dag: DAGCircuit):
+        new_dag = copy.deepcopy(dag)
+
+        for qreg in new_dag.qregs.values():
+            new_dag.apply_operation_front(Reset(), qreg, [])
+
+        return new_dag
+
+
+class QubitDrift(TransformationPass):
+    """Pass to rotate qubits during delays
+
+    This pass adds rotations that mimic the qubit being detuned from the drive
+    frequency (while assuming that the gate times are negligible; rotations are
+    only added for delays).
+    """
+
+    def __init__(self, qubit_frequencies: Sequence[float], dt: float):
+        super().__init__()
+        self.qubit_frequencies = qubit_frequencies
+        self.dt = dt
+
+    def run(self, dag: DAGCircuit):
+        qubit_indices = {bit: index for index, bit in enumerate(dag.qubits)}
+
+        new_dag = dag.copy_empty_like()
+
+        for node in dag.topological_op_nodes():
+            new_dag.apply_operation_back(node.op, node.qargs, node.cargs)
+
+            if node.name == "delay":
+                q0 = qubit_indices[node.qargs[0]]
+                if self.qubit_frequencies[q0] is None:
+                    continue
+                if node.op.unit == "dt":
+                    duration = node.op.duration * self.dt
+                elif node.op.unit != "s":
+                    duration = apply_prefix(node.op.duration, node.op.unit)
+                angle = 2 * np.pi * self.qubit_frequencies[q0] * duration
+                angle = angle % (2 * np.pi)
+                new_dag.apply_operation_back(RZGate(angle), [node.qargs[0]], [])
 
-class T2HahnBackend(BackendV1):
+        return new_dag
+
+
+class T2HahnBackend(BackendV2):
     """
     A simple and primitive backend, to be run by the T2Hahn tests
     """
 
     def __init__(
         self,
-        t2hahn=None,
-        frequency=None,
-        initialization_error=None,
-        readout0to1=None,
-        readout1to0=None,
+        t2hahn: Union[None, float, Sequence[Union[float, None]]] = None,
+        frequency: Union[None, float, Sequence[Union[float, None]]] = None,
+        initialization_error: Union[None, float, Sequence[Union[float, None]]] = None,
+        readout0to1: Union[None, float, Sequence[Union[float, None]]] = None,
+        readout1to0: Union[None, float, Sequence[Union[float, None]]] = None,
+        seed: int = 9000,
+        dt: float = 1 / 4.5e9,
+        num_qubits: Union[None, int] = None,
```

Comment:
```suggestion
        num_qubits: Optional[int] = None,
```

Reply:
Addressed in 3c49a98e7a728f9f4fe4f3312300444c1a2dcd0c